_posts/2020-01-24-inception-pipeline-cdk.md

```diff
@@ -0,0 +1,395 @@
+});
+new ReactFrontEndStack(app, "NonProdReactFrontEndStack", { env: nonProdAccount });
+new ReactFrontEndStack(app, "ProdReactFrontEndStack", { env: ProdAccount });
+```
+
+As you can see, I've got a build account which is where the pipeline lives, I've got a non-prod and
+prod account which both reference the same stack. Neato! Just by passing environment variables in, I
+can create as many stacks as I need.
+
+As an aside, if I wanted to test my stack in a new region for example, I could easily spin one up
+like so:
+
+```typescript
+const nonProdAccountIreland = { account: "456", region: "eu-west-1" };
+new ReactFrontEndStack(app, "NonProdReactFrontEndStackIreland", { env: nonProdAccountIreland });
+```
+
+So now that we've specified a destination account, let's update our pipeline by creating a new
+stage. At the top of the pipeline file, I've also defined the stack name for reasons you'll see
+soon:
+
+```typescript
+const NPRFES = "NonProdReactFrontEndStack";
+
+{
+  stageName: "Non-Production",
+  actions: [
+    new codepipeline_actions.CloudFormationCreateUpdateStackAction({
+      // NOTE: Here is the account number of the destination account.
+      account: props.nonProd.env.account,
+      actionName: "DeployFrontEndInfrastructure",
+      adminPermissions: true,
+      // Hard coded as cdk only synthesises the template. The deployment is separate.
+      templatePath: cdkBuildOutput.atPath(`${NPRFES}.template.json`),
+      stackName: `${NPRFES}`,
+      region: props.nonProd.env.region,
+      runOrder: 1,
+    }),
+    new codepipeline_actions.CodeBuildAction({
+      actionName: "BuildReactApp",
+      project: buildDeployReactNonProd,
+      input: sourceOutput,
+      runOrder: 2,
+    }),
+  ],
+},
+```
+
+But where is that template coming from? Glad you asked. It's part of the synthesised template which
+we generate the build cdk stage of the pipeline. The artifacts array needs to be updated:
+
+```typescript
+artifacts: {
+  "base-directory": "cdj-build-pipeline/dist",
+  files: [
+    `${this.stackName}.template.json`,
+    `${CRSS}.template.json`,
+    `${NPRFES}.template.json`,
+  ],
+},
+```
+
+The eagle-eyed reader among us will notice some strange string substitution going on here, let me
+explain. When the CDK synthesises it will generate stacks for everything that is required. This
+includes sub-stacks (our front-end stack, for example), as well as supporting stacks that are needed
+in destination accounts in order for the deployments to work.
+
+```typescript
+const NPRFES = "NonProdReactFrontEndStack";
+const CASS = "cross-account-support-stack-<account number>";
+```
+
+The CASS variable I've invented above is the name I've given to the template that is generated by a
+synthesis of the CDK application we're working with.
+{:refdef: style="text-align: center;"}
+<img src="/img/cdk_synth_out.png">
+{: refdef}
+
+In order to facilitate the CloudFormation step which would deploy infrastructure in our destination
```

Comment:
This didn't read particularly cleanly to me... Maybe something of the form:
In order to allow the pipeline to deploy cross-account, we need to provision a role and permissions for CloudFormation to assume. We do this through the intermediate step in the BuildAndAdministerPipeline stage.